### ISSUE TITLE
Switch to trusted publishing for PyPI/TestPyPI

### ISSUE THREAD

**Issue by woodruffw** [dateutil/dateutil#1298]
Hi there!

I'm going through the top/critical PyPI packages to see which ones aren't using [trusted publishing](https://docs.pypi.org/trusted-publishers/), and I noticed that `dateutil` isn't 🙂 

Trusted publishing is supported on both PyPI and TestPyPI and it looks like your current publishing workflow is entirely GitHub-based, so switching away from a manual API token should be straightforward (just a matter of configuring the publisher on PyPI and TestPyPI and tweaking the current publishing workflow slightly).

If you're interested, I'm happy to send a PR to tweak your publishing workflow to make it use a trusted publisher. Otherwise, please feel free to close this!

**Comment by pganssle** [comment]
If I understand correctly, this is basically the same as what I'm doing now (giving Github authority to act on my behalf with a specific scope on PyPI), except that now PyPI knows that it's GitHub is the one with the delegated authority and they have to prove that they are GitHub before using that authority?

If so, it sounds like there are no real trade-offs and trusted publishing is strictly superior. If you're still interested in sending a PR, I'd be happy to accept it. Otherwise I suppose I'll get around to configuring this myself some time before the next release.

**Comment by woodruffw** [comment]
> If I understand correctly, this is basically the same as what I'm doing now (giving Github authority to act on my behalf with a specific scope on PyPI), except that now PyPI knows that it's GitHub is the one with the delegated authority and they have to prove that they are GitHub before using that authority?

Yep, that understanding is correct -- trusted publishing uses API tokens under the hood, but only gives one to the CI/CD provider (GitHub Actions) if it can prove that it's the delegated authority for the given PyPI project.

And great! I'll try and send a PR tonight.

**Comment by woodruffw** [comment]
Much delayed, but I've opened #1375 for this.

**Comment by lmmx** [comment]
Would like to express my support for this in the wake of the litellm supply chain attack<sup>*</sup>, python-dateutil is one of the most downloaded packages on PyPI ([by downloads](https://hugovk.github.io/top-pypi-packages/)) yet to use the pypa/gh-action-pypi-publish action (which supports Trusted Publishing). I see there is already a twine workflow, so this is fairly 'low hanging fruit' that makes the PyPI publishing token unavailable to exfiltration attacks.

Instead, GitHub vouches for the identity, and you pre-declare a particular CI workflow (by filename). It also requires creating an 'environment' for the repo on GitHub (suggested name "pypi") which you also register the package with on PyPI.

<sup>*</sup> Post-mortems of the litellm attack: 
1. https://futuresearch.ai/blog/litellm-pypi-supply-chain-attack/
2. https://snyk.io/articles/poisoned-security-scanner-backdooring-litellm/
